File: android/app/src/main/kotlin/com/ophymx/chapbook/app/ui/ReaderScreen.kt
package com.ophymx.chapbook.app.ui

import android.app.Application
import android.content.ActivityNotFoundException
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.activity.compose.BackHandler
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.List
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.BottomAppBar
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalLifecycleOwner
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.ophymx.chapbook.Theme
import com.ophymx.chapbook.app.R
import com.ophymx.chapbook.app.container
import com.ophymx.chapbook.app.model.ReaderState
import com.ophymx.chapbook.app.model.ReaderViewModel
import kotlinx.coroutines.launch

/** Which sheet is up, if any. */
enum class Sheet { Contents, Search, Marks, Settings }

/** A tap on a stored highlight: which, and where on the page view. */
data class HighlightTap(val id: Long, val xPx: Float, val yPx: Float)

/** The page's ground, so the safe-inset strips match the engine's theme. */
@Composable
private fun paper(theme: Theme?): Color = when (theme) {
    Theme.LIGHT -> Color.White
    Theme.SEPIA -> Color(0xFFF6F0E2)
    Theme.DARK -> Color(0xFF121212)
    null -> MaterialTheme.colorScheme.background
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReaderScreen(bookId: Long, onBack: () -> Unit) {
    val context = LocalContext.current
    val container = context.container
    val vm: ReaderViewModel = viewModel(
        key = "reader-$bookId",
        factory = ReaderViewModel.factory(context.applicationContext as Application, container, bookId),
    )
    val state by vm.state.collectAsStateWithLifecycle()
    val place by vm.place.collectAsStateWithLifecycle()
    val settings by vm.settings.collectAsStateWithLifecycle()
    val fontFamily by vm.fontFamily.collectAsStateWithLifecycle()
    val marks by vm.marks.collectAsStateWithLifecycle()
    val search by vm.search.collectAsStateWithLifecycle()
    val selection by vm.selection.collectAsStateWithLifecycle()

    var chrome by remember { mutableStateOf(false) }
    var sheet by remember { mutableStateOf<Sheet?>(null) }
    var noteDialog by remember { mutableStateOf(false) }
    var highlightTap by remember { mutableStateOf<HighlightTap?>(null) }
    var selectionBounds by remember { mutableStateOf<SelectionBounds?>(null) }
    var failed by remember { mutableStateOf<Pair<Int, String>?>(null) }
    var unopenable by remember { mutableStateOf<String?>(null) }
    val snackbar = remember { SnackbarHostState() }
    val scope = rememberCoroutineScope()
    val copied = stringResource(R.string.copied)

    // Back peels one layer at a time: a sheet, then a selection, then the book.
    BackHandler {
        when {
            sheet != null -> sheet = null
            selection != null -> vm.clearSelection()
            chrome -> chrome = false
            else -> onBack()
        }
    }

    // `onStop` is the last callback Android guarantees: save while there
    // is a process to save from.
    val lifecycle = LocalLifecycleOwner.current.lifecycle
    DisposableEffect(lifecycle) {
        val observer = LifecycleEventObserver { _, event ->
            if (event == Lifecycle.Event.ON_STOP) vm.stopped()
        }
        lifecycle.addObserver(observer)
        onDispose { lifecycle.removeObserver(observer) }
    }

    failed?.let { (spine, message) ->
        val text = stringResource(R.string.page_failed, spine + 1, message)
        LaunchedEffect(spine, message) {
            snackbar.showSnackbar(text)
            failed = null
        }
    }
    unopenable?.let { href ->
        val text = stringResource(R.string.link_failed, href)
        LaunchedEffect(href) {
            snackbar.showSnackbar(text)
            unopenable = null
        }
    }

    // The page never goes under the status bar, the camera cutout or the
    // gesture bar: the paper colour fills those, and the page box is
    // what is left. The chrome overlays the page, inside the same bounds.
    Box(
        Modifier
            .fillMaxSize()
            .background(paper(settings?.theme))
            .safeDrawingPadding(),
    ) {
        when (val s = state) {
            ReaderState.Opening -> CircularProgressIndicator(Modifier.align(Alignment.Center))
            ReaderState.Gone -> Column(
                Modifier.align(Alignment.Center).padding(32.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(stringResource(R.string.open_gone), style = MaterialTheme.typography.bodyLarge)
                TextButton(onClick = onBack) { Text(stringResource(R.string.back)) }
            }
            is ReaderState.Reading -> {
                var page by remember { mutableStateOf<PageView?>(null) }
                AndroidView(
                    factory = { ctx ->
                        PageView(ctx, s.session).apply {
                            onMoved = vm::moved
                            onMenu = { chrome = !chrome }
                            onPageFailed = { spine, message -> failed = spine to message }
                            onExternalLink = { href -> if (!openExternal(ctx, href)) unopenable = href }
                            onHighlightTapped = { id, x, y -> highlightTap = HighlightTap(id, x, y) }
                            onSelection = { bounds ->
                                selectionBounds = bounds
                                if (bounds == null) vm.clearSelection() else vm.selected(bounds.start, bounds.end)
                            }
                            page = this
                        }
                    },
                    modifier = Modifier.fillMaxSize(),
                )
                // Volume keys arrive at the activity, and the model's
                // changes want a repaint; both reach this page while it
                // is the one showing.
                DisposableEffect(page) {
                    val view = page
                    container.keys.onKeyDown = view?.let { v -> v::handleKey }
                    container.keys.bindsKey = view?.let { v -> v::bindsKey }
                    vm.onNeedsRedraw = view?.let { v -> { v.invalidate() } }
                    onDispose {
                        container.keys.onKeyDown = null
                        container.keys.bindsKey = null
                        vm.onNeedsRedraw = null
                    }
                }

                selectionBounds?.let { bounds ->
                    SelectionBar(
                        bounds = bounds,
                        onHighlight = vm::highlightSelection,
                        onNote = { noteDialog = true },
                        onCopy = {
                            val text = selection?.text ?: ""
                            context.getSystemService(ClipboardManager::class.java)
                                ?.setPrimaryClip(ClipData.newPlainText("chapbook", text))
                            vm.clearSelection()
                            scope.launch { snackbar.showSnackbar(copied) }
                        },
                    )
                }
                highlightTap?.let { tap ->
                    HighlightMenu(
                        tap = tap,
                        onColor = { color ->
                            vm.recolorHighlight(tap.id, color)
                            highlightTap = null
                        },
                        onRemove = {
                            vm.removeMark(tap.id)
                            highlightTap = null
                        },
                        onDismiss = { highlightTap = null },
                    )
                }

                when (sheet) {
                    Sheet.Contents -> ContentsSheet(
                        entries = s.toc,
                        currentSpine = place.spine,
                        onPick = {
                            vm.gotoToc(it)
                            sheet = null
                            chrome = false
                        },
                        onDismiss = { sheet = null },
                    )
                    Sheet.Search -> SearchSheet(
                        state = search,
                        onQuery = vm::search,
                        onPick = {
                            vm.gotoHit(it)
                            sheet = null
                            chrome = false
                        },
                        onDismiss = { sheet = null },
                    )
                    Sheet.Marks -> MarksSheet(
                        marks = marks,
                        onBookmark = vm::addBookmark,
                        onPick = {
                            vm.gotoMark(it.id)
                            sheet = null
                            chrome = false
                        },
                        onDelete = { vm.removeMark(it.id) },
                        onDismiss = { sheet = null },
                    )
                    Sheet.Settings -> SettingsSheet(
                        settings = settings,
                        fontFamily = fontFamily,
                        families = s.fontFamilies,
                        onSettings = vm::applySettings,
                        onFamily = vm::setFontFamily,
                        onReset = vm::resetBookSettings,
                        onDismiss = { sheet = null },
                    )
                    null -> {}
                }
                if (noteDialog) {
                    NoteDialog(
                        onSave = {
                            vm.noteOnSelection(it)
                            noteDialog = false
                        },
                        onDismiss = { noteDialog = false },
                    )
                }
            }
        }

        val bars = TopAppBarDefaults.topAppBarColors(
            containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.94f),
        )
        AnimatedVisibility(visible = chrome, modifier = Modifier.align(Alignment.TopCenter)) {
            TopAppBar(
                colors = bars,
                title = {
                    Column {
                        Text(place.title, style = MaterialTheme.typography.titleMedium, maxLines = 1)
                        Text(
                            stringResource(
                                R.string.reader_progress,
                                place.spine + 1,
                                place.spineLen,
                                place.page + 1,
                                place.pageCount,
                            ),
                            style = MaterialTheme.typography.labelSmall,
                        )
                    }
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = stringResource(R.string.back))
                    }
                },
                actions = {
                    // The engine's Back: where the reader was before the
                    // last link, greyed out by absence rather than state.
                    if (place.canGoBack) {
                        TextButton(onClick = vm::goBack) { Text(stringResource(R.string.return_back)) }
                    }
                },
            )
        }
        AnimatedVisibility(visible = chrome, modifier = Modifier.align(Alignment.BottomCenter)) {
            BottomAppBar(containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.94f)) {
                IconButton(onClick = { sheet = Sheet.Contents }) {
                    Icon(Icons.AutoMirrored.Filled.List, contentDescription = stringResource(R.string.contents))
                }
                IconButton(onClick = { sheet = Sheet.Search }) {
                    Icon(Icons.Default.Search, contentDescription = stringResource(R.string.search))
                }
                IconButton(onClick = { sheet = Sheet.Marks }) {
                    Icon(Icons.Default.Star, contentDescription = stringResource(R.string.marks))
                }
                IconButton(onClick = { sheet = Sheet.Settings }) {
                    Icon(Icons.Default.Settings, contentDescription = stringResource(R.string.settings))
                }
            }
        }
        SnackbarHost(snackbar, Modifier.align(Alignment.BottomCenter))
    }
}

/** An `http(s)` link the engine will not follow. False when no app takes it. */
private fun openExternal(context: Context, href: String): Boolean = try {
    context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(href)))
    true
} catch (e: ActivityNotFoundException) {
    false
}

File: ios/App/Sources/ChapbookApp/ReaderScreen.swift
import Chapbook
import ChapbookAppModel
import SwiftUI

/// Which sheet is up, if any.
enum Sheet: String, Identifiable {
    case contents, search, marks, settings
    var id: String { rawValue }
}

/// A tap on a stored highlight: which, and where on the page.
struct HighlightTap: Identifiable {
    let id: Int64
    let point: CGPoint
}

/// The page's ground, so the safe-inset strips match the engine's theme.
private func paper(_ theme: Theme?) -> Color {
    switch theme {
    case .light: Color.white
    case .sepia: Color(red: 0xF6 / 255, green: 0xF0 / 255, blue: 0xE2 / 255)
    case .dark: Color(red: 0x12 / 255, green: 0x12 / 255, blue: 0x12 / 255)
    case nil: Color(.systemBackground)
    }
}

struct ReaderScreen: View {
    @Environment(\.dismiss) private var dismiss
    @Environment(\.openURL) private var openURL
    @StateObject private var vm: ReaderViewModel

    @State private var chrome = false
    @State private var sheet: Sheet?
    @State private var noteDialog = false
    @State private var noteBody = ""
    @State private var highlightTap: HighlightTap?
    @State private var selectionBounds: SelectionBounds?
    @State private var toast: String?
    @State private var page: PageView?

    init(container: AppContainer, bookID: Int64) {
        _vm = StateObject(
            wrappedValue: ReaderViewModel(
                bookID: bookID, shelf: container.shelf, opener: container.opener,
                preferences: container.preferences))
    }

    var body: some View {
        ZStack {
            // The page never goes under the status bar, the notch or the
            // home indicator: the paper colour fills those, and the page
            // box is what is left. The chrome overlays the page, inside
            // the same bounds.
            paper(vm.settings?.theme).ignoresSafeArea()
            switch vm.state {
            case .opening:
                ProgressView()
            case .gone:
                ContentUnavailableView {
                    Label(L("open_gone"), systemImage: "book.closed")
                } actions: {
                    Button(L("back")) { dismiss() }
                }
            case .reading(let reading):
                reader(reading)
            }
            chromeBars
            if let toast {
                Text(toast)
                    .font(.footnote)
                    .padding(10)
                    .background(.regularMaterial, in: RoundedRectangle(cornerRadius: 8))
                    .frame(maxHeight: .infinity, alignment: .bottom)
                    .padding(.bottom, chrome ? 72 : 24)
                    .task { try? await Task.sleep(nanoseconds: 3_000_000_000); self.toast = nil }
            }
        }
        .toolbar(.hidden, for: .navigationBar)
        .navigationBarBackButtonHidden(true)
        .statusBarHidden(!chrome)
        // `didEnterBackground` is the last callback the platform
        // guarantees: save while there is a process to save from.
        .onReceive(NotificationCenter.default.publisher(for: UIApplication.didEnterBackgroundNotification)) { _ in
            vm.suspend()
        }
        .onReceive(NotificationCenter.default.publisher(for: UIApplication.didReceiveMemoryWarningNotification)) { _ in
            vm.memoryWarning()
        }
        .onDisappear { vm.close() }
    }

    @ViewBuilder
    private func reader(_ reading: Reading) -> some View {
        PageViewRepresentable(
            session: reading.session, kind: reading.kind,
            onMoved: { vm.moved($0) },
            onMenu: { chrome.toggle() },
            onPageFailed: { spine, message in toast = L("page_failed", spine + 1, message) },
            onExternalLink: { href in
                if let url = URL(string: href) { openURL(url) } else { toast = L("link_failed", href) }
            },
            onHighlightTapped: { id, point in highlightTap = HighlightTap(id: id, point: point) },
            onSelection: { bounds in
                selectionBounds = bounds
                vm.selected(bounds?.locators)
            },
            install: { view in
                page = view
                vm.onNeedsRedraw = { [weak view] in view?.render() }
            }
        )
        .overlay(alignment: .topLeading) {
            if let bounds = selectionBounds {
                SelectionBar(
                    bounds: bounds,
                    onHighlight: { vm.highlightSelection() },
                    onNote: { noteDialog = true },
                    onCopy: {
                        UIPasteboard.general.string = vm.selection?.text ?? ""
                        vm.clearSelection()
                        toast = L("copied")
                    })
            }
        }
        .confirmationDialog(
            L("mark_highlight"), isPresented: Binding(get: { highlightTap != nil }, set: { if !$0 { highlightTap = nil } })
        ) {
            if let tap = highlightTap {
                Button(L("color_theme")) { vm.recolorHighlight(tap.id, color: nil) }
                ForEach(highlightColors, id: \.1) { label, color in
                    Button(L(label)) { vm.recolorHighlight(tap.id, color: color) }
                }
                Button(L("remove_highlight"), role: .destructive) { vm.removeMark(tap.id) }
            }
        }
        .alert(L("note_title"), isPresented: $noteDialog) {
            TextField(L("note_hint"), text: $noteBody)
            Button(L("save")) {
                let body = noteBody
                noteBody = ""
                if !body.trimmingCharacters(in: .whitespaces).isEmpty { vm.note(onSelection: body) }
            }
            Button(L("cancel"), role: .cancel) { noteBody = "" }
        }
        .sheet(item: $sheet) { which in
            Group {
                switch which {
                case .contents:
                    ContentsSheet(entries: reading.contents, currentSpine: vm.place.spine) { entry in
                        vm.go(to: entry)
                        sheet = nil
                        chrome = false
                    }
                case .search:
                    SearchSheet(vm: vm) { hit in
                        vm.go(toHit: hit)
                        sheet = nil
                        chrome = false
                    }
                case .marks:
                    MarksSheet(vm: vm) { mark in
                        vm.go(toMark: mark.id)
                        sheet = nil
                        chrome = false
                    }
                case .settings:
                    SettingsSheet(vm: vm, families: reading.fontFamilies)
                }
            }
            .presentationDetents([.medium, .large])
        }
    }

    private var chromeBars: some View {
        VStack {
            if chrome {
                HStack(spacing: 12) {
                    Button { dismiss() } label: { Image(systemName: "chevron.backward") }
                    VStack(alignment: .leading, spacing: 4) {
                        Text(vm.place.title).font(.headline).lineLimit(1)
                        ProgressReadout(place: vm.place, preferences: vm.preferences)
                    }
                    Spacer()
                    // The engine's Back: where the reader was before the
                    // last link, greyed out by absence rather than state.
                    if vm.place.canGoBack {
                        Button(L("return_back"), action: { vm.goBack() })
                    }
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 10)
                .background(.regularMaterial)
                .transition(.move(edge: .top).combined(with: .opacity))
            }
            Spacer()
            if chrome {
                HStack {
                    Spacer()
                    Button { sheet = .contents } label: { Label(L("contents"), systemImage: "list.bullet") }
                    Spacer()
                    Button { sheet = .search } label: { Label(L("search"), systemImage: "magnifyingglass") }
                    Spacer()
                    Button { sheet = .marks } label: { Label(L("marks"), systemImage: "bookmark") }
                    Spacer()
                    Button { sheet = .settings } label: { Label(L("settings"), systemImage: "textformat.size") }
                    Spacer()
                }
                .labelStyle(.iconOnly)
                .font(.title3)
                .padding(.vertical, 14)
                .background(.regularMaterial)
                .transition(.move(edge: .bottom).combined(with: .opacity))
            }
        }
        .animation(.easeInOut(duration: 0.15), value: chrome)
    }
}

/// The whole-book bar, and a readout whose words are the reader's choice.
private struct ProgressReadout: View {
    let place: ChapbookAppModel.Place
    @ObservedObject var preferences: Preferences

    private var text: String {
        switch preferences.progressLabel {
        case .percent:
            L("progress_percent", Int((place.bookFraction * 100).rounded()))
        case .pagesLeft:
            {
                let left = max(0, place.pageCount - place.page - 1)
                return left == 0 ? L("progress_last_page") : L("progress_pages_left", left)
            }()
        case .chapterPage:
            L("reader_progress", place.spine + 1, place.spineLength, place.page + 1, place.pageCount)
        }
    }

    var body: some View {
        HStack(spacing: 10) {
            ProgressView(value: place.bookFraction)
                .progressViewStyle(.linear)
                .frame(maxWidth: 160)
            Text(text).font(.caption).foregroundStyle(.secondary).monospacedDigit()
        }
    }
}

let highlightColors: [(String, String)] = [
    ("color_yellow", "#ffe082"), ("color_green", "#a5d6a7"), ("color_blue", "#90caf9"), ("color_pink", "#f48fb1"),
]
